### PR TITLE
Add string length validators

### PR DESCRIPTION
**Pull request by aljones15**

### CODE REVIEW

File: schemas/service.js [hunk 1]
```diff
@@ -39,10 +40,13 @@ const serviceDescriptor = {
       // FIXME: this will be validated on some const TBD, possibly
       // SeeAlsoService
       type: 'string',
+      minLength: 2,
+      maxLength: cfg.maxLength
     },
     serviceEndpoint: {
       // TODO: can this be improved? pattern starting with https://?
       type: 'string',
+      maxLength: cfg.maxLength,
```
Comment (dlongley) [inline schemas/service.js:49]: Service IDs and Service Descriptors aren't supported on Veres One, so maybe just try to remove them here. Also, we can't use configurable values for these validators -- they have to be hard coded.

File: schemas/did-uuid.js [hunk 2]
```diff
@@ -12,11 +12,15 @@ const cfg = config['veres-one-validator'];
 const prefix = cfg.environment === 'test' ? 'did:v1:test' : 'did:v1';
 /* eslint-disable-next-line max-len */
 const pattern = `^${prefix}:uuid:([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12})$`;
+const expectedLength = cfg.environment === 'test' ? 53 : 48;
 
 const schema = {
   title: 'DID UUID',
   type: 'string',
   pattern,
+  // prefix can be 11 chars or 6 chars for a total of 53 or 48
```
Comment (davidlehn) [inline schemas/did-uuid.js:21]: This comment here is confusing.  Perhaps move above the expectedLength calculation?

File: lib/constants.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// this is to prevent overly long strings
+const maxLength = 100;
```
Comment (davidlehn) [inline lib/constants.js:2]: This naming is vague.  Maybe qualified naming like "maxFooLength" for common maximums would help?  That could at least remove the oddness of "maxLength * 2" where the maxLength isn't even a max!  There are lots of varied uses below, so I'm not sure what the best pattern is here.

File: schemas/service.js (hunk 1 above)
Comment (aljones15) [reply schemas/service.js]: I removed the config maxLength and made a constant. Removing service and service descriptions is a little more involved.

File: schemas/did-uuid.js (hunk 2 above)
Comment (aljones15) [reply schemas/did-uuid.js]: I moved this earlier today. Not sure which commit it is.

File: lib/constants.js (hunk 3 above)
Comment (aljones15) [reply lib/constants.js:2]: This is general one of the problems with this PR. I have tried to removed maxLength where I could, but there are still some there.

File: schemas/did-uuid.js [hunk 4]
```diff
@@ -12,11 +12,15 @@ const cfg = config['veres-one-validator'];
 const prefix = cfg.environment === 'test' ? 'did:v1:test' : 'did:v1';
 /* eslint-disable-next-line max-len */
 const pattern = `^${prefix}:uuid:([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-4[a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12})$`;
+// prefix can be 11 chars or 6 chars for a total of 53 or 48
+const expectedLength = cfg.environment === 'test' ? 53 : 48;
```
Comment (dlongley) [inline schemas/did-uuid.js:16]:
```suggestion
const expectedLength = prefix.length + 42;
```

File: schemas/did.js [hunk 5]
```diff
@@ -18,6 +18,9 @@ const schema = {
   description: 'A decentralized identifier.',
   type: 'string',
   pattern,
+  minLength: cfg.environment === 'test' ? 17 : 12,
```
Comment (dlongley) [inline schemas/did.js:21]:
```suggestion
  minLength: prefix.length + 6,
```

Comment (dlongley) [reply schemas/did.js:21]: I see this happens in multiple places. If these are all based off of the prefix, just use `prefix.length` and add to the base value everywhere.

File: schemas/veres-one-validator.js [hunk 6]
```diff
@@ -5,12 +5,19 @@
...
 const did = require('./did');
 const didReference = require('./didReference');
 const didUuid = require('./did-uuid');
 const {serviceDescriptor, serviceId} = require('./service');
 const urnUuid = require('./urn-uuid');
 
+// can be a did or a url
+const creator = {
+  // FIXME this will allow any string with minLength 1 or more
+  anyOf: [schemas.url(), did(), didReference(), didUuid()]
```
Comment (dlongley) [inline schemas/veres-one-validator.js:18]: When is creator ever a DID uuid or a `schemas.url`? This seems too restrictive. Can we minimize this?

File: schemas/veres-one-validator.js [hunk 7]
```diff
@@ -20,12 +27,18 @@ const caveat = {
...
+      maxLength
       // FIXME: enable when term is finalized
       // enum: ['VeresOneElectorTicketAgent']
     }
   }
 };
 
+const invocationTarget = {
+  // FIXME this will allow any string with minLength 1 or more
+  anyOf: [schemas.url(), did(), didReference(), didUuid()]
```
Comment (dlongley) [inline schemas/veres-one-validator.js:39]: Why isn't this more restrictive?

**Review by dlongley — CHANGES_REQUESTED** [review]

File: schemas/veres-one-validator.js (hunk 6 above)
Comment (aljones15) [reply schemas/veres-one-validator.js:17]: It looks like this is due to a larger issue. `creator` is always a URL in all test data up to this point.

File: schemas/veres-one-validator.js (hunk 7 above)
Comment (aljones15) [reply schemas/veres-one-validator.js]: `invocationTarget` can be a `uuid`, `did` or `didUuid` now: https://github.com/veres-one/veres-one-validator/pull/38/commits/18dc092996069267c1b807001b1aa31f43402744

File: schemas/veres-one-validator.js (hunk 6 above)
Comment (dlongley) [reply schemas/veres-one-validator.js:17]: I'll add a punchlist item for ensuring veres-one validators only accept specific URL types.

File: schemas/did.js [hunk 8]
```diff
@@ -8,16 +8,19 @@ const {config} = bedrock;
 require('../lib/config');
 
 const cfg = config['veres-one-validator'];
-
-const pattern = cfg.environment === 'test' ?
-  '^(did\:v1\:test\:nym\:)([-_A-Za-z0-9.]+)$' :
-  '^(did\:v1\:nym\:)([-_A-Za-z0-9.]+)$';
+const prefix = cfg.environment === 'test' ? 'did:v1:test' : 'did:v1';
```
Comment (dlongley) [inline schemas/did.js:11]: I'm going to accept this -- but we're going to have to expand this to accept more than just the `test` prefix in the future, so we can support multiple test nets.

File: schemas/veres-one-validator.js [hunk 9]
```diff
@@ -440,15 +447,24 @@ const baseCapability = {
...
     },
     created: schemas.w3cDateTime(),
-    creator: {type: 'string'},
-    invocationTarget: {type: 'string'},
-    proofValue: {type: 'string'},
+    invocationTarget,
+    proofValue: {
+      type: 'string',
+      // this should be the base58 representation of a 512 bit hash.
+      // this is max 89 characters
```
Comment (dlongley) [inline schemas/veres-one-validator.js:464]:
```suggestion
      // this should be the multibase base58 representation of a 64-byte
      // ed25519 signature value; this is max 89 characters
```

File: schemas/veres-one-validator.js [hunk 10]
```diff
@@ -117,21 +125,21 @@ const publicKey = {
...
     type: {
       type: 'string',
       enum: ['Ed25519VerificationKey2020', 'X25519KeyAgreementKey2020'],
     },
     controller: did(),
-    // FIXME: make schema require this for Ed25519VerificationKey2020
     publicKeyMultibase: {
-      type: 'string'
+      type: 'string',
+      // base58 encoding of 32 byte seed material
```
Comment (dlongley) [inline schemas/veres-one-validator.js:141]:
```suggestion
      // multibase base58 encoding of multicodec public key
```

File: schemas/did.js [hunk 11]
```diff
@@ -8,16 +8,19 @@ const {config} = bedrock;
...
+// prefix + :nym:
+const minLength = prefix.length + 5;
 
 const schema = {
   title: 'Decentralized Identifier',
   description: 'A decentralized identifier.',
   type: 'string',
   pattern,
+  minLength,
+  // prefix plus a single base58 encoded 32 byte buffer
```
Comment (dlongley) [inline schemas/did.js:22]:
```suggestion
  // prefix plus a single multibase base58 encoded multicodec public key
```

**Review by dlongley — APPROVED** [review]
Approving as a good step forward toward more restrict validators. Please apply all of my suggestions -- the comments are inaccurate without them.

**Comment by msporny** [comment]
Merged witness pool upgrade into this branch, all tests continue to work, merging this into v1-dingo-rc1.

**Pull request merged**